#!/usr/bin/env Rscript

###################################################
# This R script takes an input two directories of
# .fastq.gz files, corresponding to matched forward
# and reverse sequence files,
# and outputs a tsv file of the dada2 processed sequence
# table. It is intended for use with the QIIME2 plugin
# for DADA2.
#
# Rscript run_dada_paired.R input_dirF input_dirR output.tsv filtered_dirF filtered_dirR 240 160 0 0 2.0 2 pooled 1.0 0 100000
####################################################

####################################################
#             DESCRIPTION OF ARGUMENTS             #
####################################################
# NOTE: All numeric arguments should be zero or positive.
# NOTE: All numeric arguments save maxEE are expected to be integers.
# NOTE: Currently the filterered_dirF/R must already exist.
# NOTE: ALL ARGUMENTS ARE POSITIONAL!
#
### FILE SYSTEM ARGUMENTS ###
#
# 1) File path to directory with the FORWARD .fastq.gz files to be processed.
#    Ex: path/to/dir/with/FWD_fastqgzs
#
# 2) File path to directory with the REVERSE .fastq.gz files to be processed.
#    Ex: path/to/dir/with/REV_fastqgzs
#
# 3) File path to output tsv file. If already exists, will be overwritten.
#    Ex: path/to/output_file.tsv
#
# 4) File path to directory to write the filtered FORWARD .fastq.gz files. These files are intermediate
#               for the full workflow. Currently they remain after the script finishes. Directory must
#               already exist.
#    Ex: path/to/dir/with/FWD_fastqgzs/filtered
#
# 5) File path to directory to write the filtered REVERSE .fastq.gz files. These files are intermediate
#               for the full workflow. Currently they remain after the script finishes. Directory must
#               already exist.
#    Ex: path/to/dir/with/REV_fastqgzs/filtered
#
### FILTERING ARGUMENTS ###
#
# 6) truncLenF - The position at which to truncate forward reads. Forward reads shorter
#               than truncLenF will be discarded.
#               Special values: 0 - no truncation or length filtering.
#    Ex: 240
#
# 7) truncLenR - The position at which to truncate reverse reads. Reverse reads shorter
#               than truncLenR will be discarded.
#               Special values: 0 - no truncation or length filtering.
#    Ex: 160
#
# 8) trimLeftF - The number of nucleotides to remove from the start of
#               each forward read. Should be less than truncLenF.
#    Ex: 0
#
# 9) trimLeftR - The number of nucleotides to remove from the start of
#               each reverse read. Should be less than truncLenR.
#    Ex: 0
#
# 10) maxEE - Reads with expected errors higher than maxEE are discarded.
#               Both forward and reverse reads are independently tested.
#    Ex: 2.0
#
# 11) truncQ - Reads are truncated at the first instance of quality score truncQ.
#                If the read is then shorter than truncLen, it is discarded.
#    Ex: 2
#
### CHIMERA ARGUMENTS ###
#
# 12) chimeraMethod - The method used to remove chimeras. Valid options are:
#               none: No chimera removal is performed.
#               pooled: All reads are pooled prior to chimera detection.
#               consensus: Chimeras are detect in samples individually, and a consensus decision
#                           is made for each sequence variant.
#    Ex: consensus
#
# 13) minParentFold - The minimum abundance of potential "parents" of a sequence being
#               tested as chimeric, expressed as a fold-change versus the abundance of the sequence being
#               tested. Values should be greater than or equal to 1 (i.e. parents should be more
#               abundant than the sequence being tested).
#    Ex: 1.0
#
### SPEED ARGUMENTS ###
#
# 14) nthreads - The number of threads to use.
#                 Special values: 0 - detect available and use all.
#    Ex: 1
#
# 15) nreads_learn - The minimum number of reads to learn the error model from.
#                 Special values: 0 - Use all input reads.
#    Ex: 1000000
#

cat(R.version$version.string, "\n")
args <- commandArgs(TRUE)

inp.dirF <- args[[1]]
inp.dirR <- args[[2]]
out.path <- args[[3]]
filtered.dirF <- args[[4]]
filtered.dirR <- args[[5]]
truncLenF <- as.integer(args[[6]])
truncLenR <- as.integer(args[[7]])
trimLeftF <- as.integer(args[[8]])
trimLeftR <- as.integer(args[[9]])
maxEE <- as.numeric(args[[10]])
truncQ <- as.integer(args[[11]])
chimeraMethod <- args[[12]]
minParentFold <- as.numeric(args[[13]])
nthreads <- as.integer(args[[14]])
nreads.learn <- as.integer(args[[15]])
trace.dirF <- args[[16]]
trace.dirR <- args[[17]]
errQuit <- function(mesg, status=1) {
  message("Error: ", mesg)
  q(status=status)
}

### VALIDATE ARGUMENTS ###

# Input directory is expected to contain .fastq.gz file(s)
# that have not yet been filtered and globally trimmed
# to the same length.
if(!(dir.exists(inp.dirF) && dir.exists(inp.dirR))) {
  errQuit("Input directory does not exist.")
} else {
  unfiltsF <- list.files(inp.dirF, pattern=".fastq.gz$", full.names=TRUE)
  unfiltsR <- list.files(inp.dirR, pattern=".fastq.gz$", full.names=TRUE)
  if(length(unfiltsF) == 0) {
    errQuit("No input forward files with the expected filename format found.")
  }
  if(length(unfiltsR) == 0) {
    errQuit("No input reverse files with the expected filename format found.")
  }
  if(length(unfiltsF) != length(unfiltsR)) {
    errQuit("Different numbers of forward and reverse .fastq.gz files.")
  }
}

# Output path is to be a filename (not a directory) and is to be
# removed and replaced if already present.
if(dir.exists(out.path)) {
  errQuit("Output filename is a directory.")
} else if(file.exists(out.path)) {
  invisible(file.remove(out.path))
}

# Convert nthreads to the logical/numeric expected by dada2
if(nthreads < 0) {
  errQuit("nthreads must be non-negative.")
} else if(nthreads == 0) {
  multithread <- TRUE # detect and use all
} else if(nthreads == 1) {
  multithread <- FALSE
} else {
  multithread <- nthreads
}

if(!dir.exists(trace.dirF) || !dir.exists(trace.dirR)) {
  errQuit("Trace directory does not exist.")
}

### LOAD LIBRARIES ###
suppressWarnings(library(methods))
suppressWarnings(library(dada2))
cat("DADA2 R package version:", as.character(packageVersion("dada2")), "\n")

### TRIM AND FILTER ###
cat("1) Filtering ")
filtsF <- file.path(filtered.dirF, basename(unfiltsF))
filtsR <- file.path(filtered.dirR, basename(unfiltsR))
out <- suppressWarnings(filterAndTrim(unfiltsF, filtsF, unfiltsR, filtsR,
                                      truncLen=c(truncLenF, truncLenR), trimLeft=c(trimLeftF, trimLeftR),
                                      maxEE=maxEE, truncQ=truncQ, rm.phix=TRUE, 
                                      multithread=multithread))
cat(ifelse(file.exists(filtsF), ".", "x"), sep="")
filtsF <- list.files(filtered.dirF, pattern=".fastq.gz$", full.names=TRUE)
filtsR <- list.files(filtered.dirR, pattern=".fastq.gz$", full.names=TRUE)
cat("\n")
if(length(filtsF) == 0) { # All reads were filtered out
  errQuit("No reads passed the filter (were truncLenF/R longer than the read lengths?)", status=2)
}

### LEARN ERROR RATES ###
# Dereplicate enough samples to get nreads.learn total reads
cat("2) Learning Error Rates\n")
NREADS <- 0
drpsF <- vector("list", length(filtsF))
drpsR <- vector("list", length(filtsR))
denoisedF <- rep(0, length(filtsF))
getN <- function(x) sum(getUniques(x))
for(i in seq_along(filtsF)) {
  drpsF[[i]] <- derepFastq(filtsF[[i]])
  drpsR[[i]] <- derepFastq(filtsR[[i]])
  NREADS <- NREADS + sum(drpsF[[i]]$uniques)
  if(NREADS > nreads.learn) { break }
}
# Run dada in self-consist mode on those samples
ddsF <- vector("list", length(filtsF))
ddsR <- vector("list", length(filtsR))
if(i==1){
  cat("2a) Forward Reads\n")
  ddsF[[1]] <- dada(drpsF[[1]], err=NULL, selfConsist=TRUE, multithread=multithread, VECTORIZED_ALIGNMENT=FALSE, SSE=1)
  cat("2b) Reverse Reads\n")
  ddsR[[1]] <- dada(drpsR[[1]], err=NULL, selfConsist=TRUE, multithread=multithread, VECTORIZED_ALIGNMENT=FALSE, SSE=1)
} else {
  cat("2a) Forward Reads\n")
  ddsF[1:i] <- dada(drpsF[1:i], err=NULL, selfConsist=TRUE, multithread=multithread, VECTORIZED_ALIGNMENT=FALSE, SSE=1)
  cat("2b) Reverse Reads\n")
  ddsR[1:i] <- dada(drpsR[1:i], err=NULL, selfConsist=TRUE, multithread=multithread, VECTORIZED_ALIGNMENT=FALSE, SSE=1)
}
errF <- ddsF[[1]]$err_out
errR <- ddsR[[1]]$err_out
cat("\n")

### PROCESS ALL SAMPLES ###
# Process samples used to learn error rates
mergers <- vector("list", length(filtsF))
if(i==1) { # breaks list assignment
  mergers[[1]] <- mergePairs(ddsF[[1]], drpsF[[1]], ddsR[[1]], drpsR[[1]])
  denoisedF[[1]] <- getN(ddsF[[1]])
} else {
  mergers[1:i] <- mergePairs(ddsF[1:i], drpsF[1:i], ddsR[1:i], drpsR[1:i])
  denoisedF[1:i] <- sapply(ddsF[1:i], getN)
}

# Loop over rest in streaming fashion with learned error rates
cat("3) Denoise remaining samples ")
if(i < length(filtsF)) {
  for(j in seq(i+1,length(filtsF))) {
    drpsF[[j]] <- derepFastq(filtsF[[j]])
    { sink("/dev/null"); ddsF[[j]] <- dada(drpsF[[j]], err=errF, multithread=multithread, VECTORIZED_ALIGNMENT=FALSE, SSE=1); sink(); }
    drpsR[[j]] <- derepFastq(filtsR[[j]])
    { sink("/dev/null"); ddsR[[j]] <- dada(drpsR[[j]], err=errR, multithread=multithread, VECTORIZED_ALIGNMENT=FALSE, SSE=1); sink(); }
    mergers[[j]] <- mergePairs(ddsF[[j]], drpsF[[j]], ddsR[[j]], drpsR[[j]], maxMismatch=10)
    denoisedF[[j]] <- getN(ddF)
    cat(".")
  }
}
cat("\n")

for (j in seq(1, length(filtsF))) {
  map_path <- file.path(trace.dirF, gsub('fastq.gz', 'dada.map', basename(filtsF[[j]])))
  uniques <- getSequences(drpsF[[j]])
  svs <- names(ddsF[[j]]$denoised[unname(ddsF[[j]]$map)])
  write.table(t(rbind(uniques, svs)),
              map_path, sep="\t", quote=FALSE, row.names=FALSE, col.names=FALSE)

  map_path <- file.path(trace.dirF, gsub('fastq.gz', 'merge.map', basename(filtsF[[j]])))
  merged <- getSequences(mergers[[j]])
  svs <- names(ddsF[[j]]$denoised[unname(mergers[[j]]$forward)])
  write.table(t(rbind(svs, merged)),
              map_path, sep="\t", quote=FALSE, row.names=FALSE, col.names=FALSE)

  map_path <- file.path(trace.dirR, gsub('fastq.gz', 'dada.map', basename(filtsR[[j]])))
  uniques <- getSequences(drpsR[[j]])
  svs <- names(ddsR[[j]]$denoised[unname(ddsR[[j]]$map)])
  write.table(t(rbind(uniques, svs)),
              map_path, sep="\t", quote=FALSE, row.names=FALSE, col.names=FALSE)

  map_path <- file.path(trace.dirR, gsub('fastq.gz', 'merge.map', basename(filtsR[[j]])))
  svs <- names(ddsR[[j]]$denoised[unname(mergers[[j]]$reverse)])
  write.table(t(rbind(svs, merged)),
              map_path, sep="\t", quote=FALSE, row.names=FALSE, col.names=FALSE)

  rm(uniques); rm(merged); rm(svs)
}
rm(drpsF); rm(drpsR); rm(ddsF); rm(ddsR)

# Make sequence table
seqtab <- makeSequenceTable(mergers)

# Remove chimeras
cat("4) Remove chimeras (method = ", chimeraMethod, ")\n", sep="")
if(chimeraMethod %in% c("pooled", "consensus")) {
  seqtab.nochim <- removeBimeraDenovo(seqtab, method=chimeraMethod, minFoldParentOverAbundance=minParentFold, multithread=multithread)
} else { # No chimera removal, copy seqtab to seqtab.nochim
  seqtab.nochim <- seqtab
}

### REPORT READ COUNTS AT EACH PROCESSING STEP ###
# Handle edge cases: Samples lost in filtering; One sample
track <- cbind(out, matrix(0, nrow=nrow(out), ncol=3))
colnames(track) <- c("input", "filtered", "denoised", "merged", "non-chimeric")
passed.filtering <- track[,"filtered"] > 0
track[passed.filtering,"denoised"] <- denoisedF
track[passed.filtering,"merged"] <- rowSums(seqtab)
track[passed.filtering,"non-chimeric"] <- rowSums(seqtab.nochim)
head(track)
#write.table(track, out.track, sep="\t",
#            row.names=TRUE, col.names=col.names, quote=FALSE)

### WRITE OUTPUT AND QUIT ###
# Formatting as tsv plain-text sequence table table
cat("6) Write output\n")
seqtab.nochim <- t(seqtab.nochim) # QIIME has OTUs as rows
col.names <- basename(filtsF)
col.names[[1]] <- paste0("#OTU ID\t", col.names[[1]])
write.table(seqtab.nochim, out.path, sep="\t",
            row.names=TRUE, col.names=col.names, quote=FALSE)
saveRDS(seqtab.nochim, gsub("tsv", "rds", out.path)) ### TESTING

q(status=0)